### PR TITLE
Bugfix for SimplifiedLayerNormalization

### PR DESCRIPTION
**Pull request by iK1D**
This PR is to fix https://github.com/microsoft/onnxruntime/issues/12930 and https://github.com/microsoft/onnxruntime/issues/12579.

In detail:
- For CPU EP, since current impl of SimplifiedLayerNormalization doesn't support input and scale having different data types, so if the sub-graph contains Cast Op, the sub-graph will not fused, this guarantee that both inputs and output data type will be same
- For CUDA EP, add (fp16, float) support to (T,V) type constraints all combinations of fp16 and float can be supported in the impl

With the fix, the original model can be run with SimplifiedLayerNormalization, which also helps to improve the perf.

### CODE REVIEW

File: onnxruntime/test/optimizer/graph_transform_test.cc [hunk 1]
```diff
@@ -3976,7 +3978,27 @@
...
+  ASSERT_TRUE(op_to_count["SimplifiedLayerNormalization"] == 0);
+}
+
+TEST_F(GraphTransformationTests, SimplifiedLayerNormWithCastsFusionTestCudaEp) {
+  auto model_uri = MODEL_FOLDER "fusion/simplified_layer_norm_with_casts.onnx";
+  std::shared_ptr<Model> p_model;
+  ASSERT_STATUS_OK(Model::Load(model_uri, p_model, nullptr, *logger_));
+  Graph& graph = p_model->MainGraph();
+  for (auto& node : graph.Nodes()) {
+    node.SetExecutionProviderType(kCudaExecutionProvider);
```
Comment (ytaous) [inline onnxruntime/test/optimizer/graph_transform_test.cc:3995]: can we add ifdef USSE_CUDA and USE_ROCM ?

File: onnxruntime/core/optimizer/layer_norm_fusion.cc [hunk 2]
```diff
@@ -535,7 +535,12 @@ Status SimplifiedLayerNormFusion::ApplyImpl(Graph& graph, bool& modified, int gr
 
     const Node* p_pow_input_node = graph_utils::GetInputNode(pow_node, 0);
     bool has_leading_cast = false;
-    if (p_pow_input_node) {
+    // If it's not GPU EP, since the CPU impl for SimplifiedLayerNormalization doesn't support input and scale
+    // having different types for now, and it may also have conflict to InsertCastTransformer,
+    // so the sub-graph will not be fused if it contains Cast Op.
+    bool is_gpu_ep = pow_node.GetExecutionProviderType() == kCudaExecutionProvider ||
```
Comment (skottmckay) [inline onnxruntime/core/optimizer/layer_norm_fusion.cc:541]: Can you explain this change a bit more? Based on the comment I would have expected the code to look for a Cast and exit if the CPU EP was involved.

I don't quite understand why we change the first branch which seems to be about setting `has_leading_cast` and a second location which has nothing to do with `has_leading_cast`.

Comment (iK1D) [reply onnxruntime/core/optimizer/layer_norm_fusion.cc:541]: There are only 4 possible cases (x=Pow->ReduceMean->Add->Sqrt->Div, and y=Mul):
1) cast(to:float)->x->cast(to:fp16)->y : SimplifiedLayerNorm(T:fp16,V:fp16)
2) cast(to:float)->x->y : SimplifiedLayerNorm(T:fp16,V:float)
3) x->cast(to:fp16)->y : SimplifiedLayerNorm(T:float,V:fp16)
4) x->y : SimplifiedlayerNorm(T:float,V:float)

They all work for CUDA EP.

For CPU EP, we have only SimplifiedlayerNorm(T:float,V:float), so only #4 works. But if for #1 and #2, we just treat the entry cast as a normal node, means has_leading_cast=false, then for #2, we can still fuse it to "cast(to:float)->SimplifiedlayerNorm(T:float,V:float)" (just like applying #4 to the x->y after cast), so the condition for CPU EP to fuse or not is just to check if there is a cast between x and y.

File: onnxruntime/test/optimizer/graph_transform_test.cc (hunk 1 above)
Comment (iK1D) [reply onnxruntime/test/optimizer/graph_transform_test.cc:3995]: We just process the graph here and will not run it. It's OK to set any EP to the nodes in the graph logically for any kind of build. There are some other UTs in the same file doing the same thing but not using macro to limit them for CUDA or ROCm build.

**Review by ytaous — DISMISSED** [review]

File: onnxruntime/core/optimizer/layer_norm_fusion.cc (hunk 2 above)
Comment (skottmckay) [reply onnxruntime/core/optimizer/layer_norm_fusion.cc:541]: Would be great to put this excellent explanation in the comment so it's captured for the next person who works on the code.

**Review by skottmckay — DISMISSED** [review]

**Review by ytaous — APPROVED** [review]

**Pull request merged**